app/src/ui/app.tsx

```diff
@@ -653,6 +654,33 @@ export class App extends React.Component<IAppProps, IAppState> {
     }
   }
 
+  private discardChanges = (
+    repository: Repository,
+    files: ReadonlyArray<WorkingDirectoryFileChange>
+  ) => {
+    if (!repository) {
+      return null
+    }
+
+    if (this.state.askForConfirmationOnDiscardChanges) {
+      return (
+        <DiscardChanges
+          key="discard-changes"
+          repository={repository}
+          dispatcher={this.props.dispatcher}
+          files={files}
+          confirmDiscardChanges={this.state.askForConfirmationOnDiscardChanges}
+          onDismissed={this.onPopupDismissed}
+          onConfirmDiscardChangesChanged={this.onConfirmDiscardChangesChanged}
+        />
+      )
+    }
+
+    this.props.dispatcher.discardChanges(repository, files)
```

Comment:
This feels weird to me, that we'd be performing an action _or_ rendering? It seems like we should perform the discard instead of showing the popup if `this.state.askForConfirmationOnDiscardChanges` is false.